### DISCUSSION TITLE
Lesson-13: upgradeTo not found error

### DISCUSSION THREAD

**Discussion by talhaEth** [Cyfrin/foundry-full-course-cu#1053] [Q&A] [answered]
I am facing upgradeTo error in this line `proxy.upgradeTo(address(newBox));`

```
here is my code
```
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/66763322/4678730a-f859-4e6d-88cd-4ef7477619a3)

```
here is results
```
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/66763322/87834668-dad0-4709-ba72-1350c51772eb)

**Comment by Niferu** [comment]
Hello @talha073 

`upgradeTo` function comes from `UUPSUpgradeable.sol` contract. Is your Box1 contract built as below?

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/110537615/8be646d3-d69d-4c13-adc2-205bba37d501)

**Reply by talhaEth** [reply]
Hello @Neftyr 
yes, I import `UUPSUpgradeable.sol` in both boxV1 and boxV2

**Reply by Niferu** [reply]
@talha073 

have you then checked if contract under name `proxy` indeed doesnt have that function? You are probably using incorrect contract udner `proxy` that might be script issue from DevOpsTools or your `UUPSUpgradeable.sol` contract is somehow outdated and it doesnt have upgradeTo function, just ctrl + click on `UUPSUpgradeable.sol` to see its code and check if it does have `upgradeTo` function.

**Reply by Niferu** [reply, marked as answer]
I think I know where the issue is. You are just using new openzeppelin contract which is here: [link](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/master/contracts/proxy/utils/UUPSUpgradeable.sol). As you can see there is indeed no function named `upgradeTo`. Just downgrade your openzeppelin contracts package version to match course versions and you will be good :)

**Reply by talhaEth** [reply]
ahha !!
yes, there is no function name `upgradeTo` in this version. If I'll use the current version of `UUPSUpgradeable,sol` then I have to call `upgradeToAndCall`  method ?

**Reply by Niferu** [reply]
@talha073 

That is correct. You can adjust your code to match new `UUPSUpgradeable.sol` or you can just downgrade your openzeppelin package with contracts to lower version, so it will match course repo. Keep in mind that changes in new openzeppelin contracts are pretty massive as I have seen, so it could be challenging to fix all the issues, but it is your choice. Now you know where the problem lies and how to solve it, good luck! :)

**Reply by talhaEth** [reply]
I am truly grateful for your support. @Neftyr
stay blessed

**Reply by Niferu** [reply]
I'm glad I could help, cheers!

**Reply by GenevaRoc** [reply]
hello, I keep having getting this error in the nft lesson from 
here is the error ` Error: 
Compiler run failed:
Error (7576): Undeclared identifier.
  --> src/MoodNft.sol:40:14:
   |
40 |         if (!_isApprovedOrOwner(msg.sender, tokenId)) {
   |              ^^^^^^^^^^^^^^^^^^`

any idea why i keep gettingb this error? I've checked and I'm not missing any step

**Reply by Niferu** [reply]
Hi @GenevaRoc 

Please setup new topic with all details and repo link, so we can take a look at it.

**Reply by GenevaRoc** [reply]
just did

**Reply by Aishwaria8786** [reply]
Can you please tell me how to downgrade Openzeppelin? I came across the same issue.

**Reply by talhaEth** [reply]
Hello @Aishwaria8786 
you can downgrade openzeppelin contacts bu using this command.
```
forge install OpenZeppelin/openzeppelin-contracts@<version>
```

Replace <version> with the specific version you want to install. For example, to install version 4.7.1, you would run the following command:
```
forge install OpenZeppelin/openzeppelin-contracts@v4.7.1
```
